Port `VF2PostLayout` to Rust

jakelishman opened the pull request:
This fully ports the `VF2PostLayout` to Rust, matching the interface already supplied in `VF2Layout`.  In Rust space, with the new semantic matchers, the form of `VF2PostLayout` with `strict_direction=False` is exactly equal to that of `VF2Layout`, while `strict_direction=True` (i.e. exact node matching) is only a relatively small abstraction difference at the point of creating the (interaction graph, scorer) pair.

The two passes functions in Rust space have names that more accurately represent the algorithms they implement, rather than the position in the transpiler pipeline they occupy.  This is because `VF2Layout` and `VF2PostLayout` running in `strict_direction=False` are both identical from Rust space; the only difference is which property-set key gets filled in with the result.  Since Rust space does not know (yet) about the property set, they are identical.

*edit*: this is part of a PR series that makes major modifications to how we handling scoring within the VF2 algorithm: https://github.com/Qiskit/qiskit/issues/14866.

Fix #12276

eginez commented:
Hey @jakelishman thanks for working on this. I am glad Qiskit is improving here, and agree lots of space for improvements in this [direction](https://arxiv.org/abs/2508.21287) ( I am one of the authors of the paper)
Quick q, I noticed that `time_limit` is getting introduced to try to push manage the overall time limit of the algorithm, would it fix this? https://github.com/Qiskit/rustworkx/issues/1376

I am wondering if you all have thought about running a worker thread on the background to be able to manage total time more accurately

Thoughts?

eginez commented:
Some other follow up questions
- In this [PR](https://github.com/Qiskit/qiskit/pull/14856) you mention that partial mapping will now get pruned early based on some condition (presumably scoring the partial layout) I am wondering how you are avoiding pruning mappings that could have partial low scores, but globally they end up being good candidates? Is that possible?
- How are you all benchmarking these enhancements?


Anyways great work

jakelishman commented:
For `time_limit`: Qiskit's VF2 passes have always had this, but we don't use it because a) it's currently unreliable as Rustworkx's (and now Qiskit's) VF2 implemention only rarely checked back in with Python space and b) it makes the pass non-deterministic. `call_limit` correlates _very_ closely with time for a given machine at constant load, but is entirely deterministic. I will at some point see about adding simple distributive parallelism to this algorithm, and tune the frequency of synchronisations, at which point it'll be much easier to include a more accurate `time_limit` (though really I'm most bothered about responding to Python signals in a timely manner without compromising the performance).

For pruning: we only prune when a branch's score exceeds the current known best, and we arrange that the scoring function is monotonic with respect to the partial match, so we never prune out candidate solutions that could become the best. It's give-or-take a classic textbook branch-and-bound algorithm, with a precise bound. The runtime improvements from such a branch-and-bound come in two forms: when there's huge error heterogeneity, you can typically prune sizeable areas off the search space; when there's large symmetry, you reduce the complexity cost of scoring because you don't rescore common prefixes.

One reason I'd be particularly interested in your new paper considering this as a comparison is that it'd be interesting to see how your deliberate avoidance of branching at the cost of increased failure rate scales compared to a branch-and-bound like this. (Also, could you comment on which version of Qiskit and what your exact scripts were when producing the graphs?)

jakelishman commented:
> How are you all benchmarking these enhancements?

Missed this one, sorry: for `VF2PostLayout` I haven't actually included any targeted benchmarks in our tracking yet, but we have a simple suite for just `VF2Layout`.  They're in the `vf2` section at the bottom of https://qiskit.github.io/qiskit (e.g. [here's one for mapping a line of 10/20 qubits onto a square grid with log-normal distributed errors](https://qiskit.github.io/qiskit/#vf2.VF2LayoutSuite.time_grid_line) - the most recent benchmarked commit makes that do the branch-and-bound I described, and you can see that for one problem size, it gives a 200x improvement [though ofc 200x is meaningless overall because they scale differently]), and the code for them is here: https://github.com/Qiskit/qiskit/blob/main/test/benchmarks/vf2.py.

They're not meant to be completely representative of true workloads, they're meant to test the characteristics on indicative problems, or ones that are easier to reason about, just as a tracking metric, so we can tell if something introduced spurious regressions, or if a PR intended to improve performance _actually_ improves performance.

eginez commented:
Awesome thanks for sharing that info, and thanks for sharing the benchmarking graphs, I was aware of their existence

 >could you comment on which version of Qiskit and what your exact scripts were when producing the graphs?

Qiskit 1.4.2, rustworkx 0.15.1
The scripts we use for the graphs are currently private; we have been thinking about open sourcing them.

jakelishman committed:
Port `VF2PostLayout` to Rust

This fully ports the `VF2PostLayout` to Rust, matching the interface
already supplied in `VF2Layout`.  In Rust space, with the new semantic
matchers, the form of `VF2PostLayout` with `strict_direction=False` is
exactly equal to that of `VF2Layout`, while `strict_direction=True`
(i.e. exact node matching) is only a relatively small abstraction
difference at the point of creating the (interaction graph, scorer)
pair.

The two passes functions in Rust space have names that more accurately
represent the algorithms they implement, rather than the position in the
transpiler pipeline they occupy.  This is because `VF2Layout` and
`VF2PostLayout` running in `strict_direction=False` are both identical
from Rust space; the only difference is which property-set key gets
filled in with the result.  Since Rust space does not know (yet) about
the property set, they are identical.

jakelishman committed:
Fix cext naming

alexanderivrii commented on crates/cext/src/transpiler/passes/vf2.rs:
```diff
@@ -265,6 +267,29 @@ pub unsafe extern "C" fn qk_vf2_layout_configuration_set_shuffle_seed(
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_vf2_layout_configuration_score_initial(
+    config: *mut VF2LayoutConfiguration,
```
Why does the config need to be mutable here?

jakelishman replied:
the function is a setter - I fixed the name of it per the prior reviews in 3a7755e619f65738ab1410b5ad1b17de0e55713d.

alexanderivrii commented on crates/transpiler/src/passes/vf2/vf2_layout.rs:
```diff
@@ -443,6 +440,18 @@ impl<T: Default> VirtualInteractions<T> {
             .add_node(self.uncoupled.swap_remove(&q).unwrap_or_default())
     }
 }
+impl<T: ::std::fmt::Debug> VirtualInteractions<T> {
```
Why do we need here that `T` implements `Debug`?

jakelishman replied:
I think I did some debugging and had `dbg!` statements in here at some point.  Removed in bf8460ffea.

alexanderivrii commented on crates/transpiler/src/passes/vf2/vf2_layout.rs:
```diff
@@ -484,6 +493,96 @@ fn build_average_coupling_map(target: &Target, errors: &ErrorMap) -> Option<Grap
+                }
+            }
+            2 => {
+                // TODO: the Python-space version of `VF2PostLayout` in strict mode has an unusual
+                // interpretation of "global" 2q operations; it defines the operation on all 2q
+                // links that _also_ have concrete instructions.  For now, we replicate that.
```
Is this something that we are planning to change at some point?

jakelishman replied:
not _necessarily_, but possibly?  I can remove the "todo" and just leave the comment if you prefer

alexanderivrii commented on crates/transpiler/src/passes/vf2/vf2_layout.rs:
```diff
@@ -654,83 +816,90 @@ pub fn vf2_layout_pass(
+    // the rest of the graph doesn't hold in the "exact" case, so we have to have VF2 match them all
+    // at once.  For example, consider a heterogeneous target where only one qubit has `rx`
+    // available, but VF2 matches a non-`rx`-using qubit onto that one during the first step,
+    // because it isn't aware there's a necessary` rx` gate in the free list.
```
super-minor:
```suggestion
    // because it isn't aware there's a necessary `rx` gate in the free list.
```

jakelishman replied:
Done in 6146de1a50

alexanderivrii commented on crates/transpiler/src/passes/vf2/vf2_layout.rs:
```diff
@@ -654,83 +816,90 @@ pub fn vf2_layout_pass(
+    // at once.  For example, consider a heterogeneous target where only one qubit has `rx`
+    // available, but VF2 matches a non-`rx`-using qubit onto that one during the first step,
+    // because it isn't aware there's a necessary` rx` gate in the free list.
+    if interactions.uncoupled.len() > 1 {
+        // ... but if there's too many uncoupled qubits, there's a combinatorial explosion in the
+        // matching time, so we just bail out.  It would be nice to deal with this better.
+        return Ok(Vf2PassReturn::NoSolution);
+    }
```
Do we have this problem even after the new early-pruning mechanism? Hmm, yeah, I guess it is. However, I am still wondering if it makes more sense to eagerly give up rather than performing the allowed `call_limit` number of calls. After all, it is also possible that we will find a better solution.

jakelishman replied:
the initial limit in `call_limit` is typically set rather high, so it can totally blow out the time of the algorithm with no real benefit. This would be nice to revisit, but it's possibly better as follow-on research?

alexanderivrii replied:
Agreed, it's probably best not to change the default parameters right now. However, it does feel strange to apply such "hacks" when there is a very natural `call_limit` option that is specifically there to bound the running time of the algorithm.

alexanderivrii commented on crates/transpiler/src/passes/vf2/vf2_layout.rs:
```diff
@@ -654,83 +816,90 @@ pub fn vf2_layout_pass(
-            if !error.is_nan() {
-                (1. - error).powi(*gate_count)
+#[pyo3(signature = (dag, target, config))]
+pub fn vf2_layout_pass_exact(
+    dag: &DAGCircuit,
+    target: &Target,
+    config: &Vf2PassConfiguration,
+) -> PyResult<Vf2PassReturn> {
```
Nice! This function is indeed quite similar to `vf2_layout_pass_average`, modulo how interaction graph is constructed.

alexanderivrii commented on qiskit/transpiler/passes/layout/vf2_utils.py:
```diff
@@ -12,31 +12,18 @@
-    return layout
 
 
+# This function is (unfortunately) used by non-VF2 places in Qiskit.
```
Only by a test in `test_star_prerouting`, right? If we really need this function there, we could just move it there, and get rid of it here.

jakelishman replied:
I think it was used more at the time I wrote the commit, but since it's no longer used, I removed it in 1859960967.  I just rewrote the simple `StarPreroutingTest` to extract the edges without all the extra baggage.

alexanderivrii commented on qiskit/transpiler/passes/layout/vf2_utils.py:
```diff
@@ -12,31 +12,18 @@
-import numpy as np
 from rustworkx import PyDiGraph, PyGraph, connected_components, weakly_connected_components
 
+from qiskit._accelerate.error_map import ErrorMap
```
Do we need to support exporting `ErrorMap` from `vf2_utils`? I guess the question here if (backward-compatibility issues aside) it would make sense to export `ErrorMap`from the parent `layout` directory instead?

jakelishman replied:
We implicitly documented it in the release note that added it, unfortunately.  I've added a comment in 1859960967.

For a follow-up, we could find a better export location - even `qiskit.transpiler.passes.layout` isn't public.

alexanderivrii commented on qiskit/transpiler/passes/layout/vf2_post_layout.py:
```diff
@@ -143,249 +129,40 @@ def run(self, dag):
         """run the layout method"""
         if self.target is None:
             raise TranspilerError("A target must be specified or a coupling map must be provided")
-        if not self.strict_direction:
-            self.avg_error_map = self.property_set["vf2_avg_error_map"]
-            if self.avg_error_map is None:
-                self.avg_error_map = vf2_utils.build_average_error_map(self.target, None)
```
This comment is intended for the lines above these, where it says:
```python
if self.target is None:
        raise TranspilerError("A target must be specified or a coupling map must be provided")
```
This pass does not accept the `coupling_map` argument! 😄 

Modulo that, the pass has the same arguments as `VF2Layout`, but somehow in a different order.

jakelishman replied:
yeah, don't blame me for that one lol.  Fixed the exception message in df6d109671.

alexanderivrii commented on test/python/compiler/test_transpiler.py:
```diff
@@ -2271,14 +2271,18 @@ def test_custom_dt_preserves_properties(self):
             basis_gates=["id", "sx", "x", "cx", "rz"],
             coupling_map=coupling_map,
-            seed=0,
+            seed=123,
```
What was the reason for changing the seed here?

jakelishman replied:
This test requires that the "no error" transpilation doesn't accidentally choose the best qubits by itself so that it can verify there's a difference.  There's so few qubits and possibilities that it's actually quite likely for that to happen, and it just so happened that my RNG changes caused it to trigger.

alexanderivrii commented on test/python/transpiler/test_vf2_post_layout.py:
```diff
@@ -379,52 +369,46 @@ def test_complete_layout_with_one_uncoupled_qubit(self, seed):
+        qc = QuantumCircuit(num_qubits)
+        qc.x([0, 1, 2, 3])
+        qc.rx(0.5, 4)
+        for pair in [(3, 2), (2, 1), (1, 0)]:
+            qc.cx(*pair)
```
This test produces a layout only because there is only a single free qubit; removing say the `(1, 0)` interaction leads to layout suddenly becoming `None`. Should we relax/remove the condition `if interactions.uncoupled.len() > 1 { return Ok(Vf2PassReturn::NoSolution); }`?

jakelishman replied:
Well, I mean I deliberately wrote the test so that it has a single free qubit - the comment above these lines says that.

Relaxing that line you've suggested causes factorial blow-up in the runtime.  It would be nice to improve the algorithm, but right now, doing so ends up with unacceptable performance - it's better to not find a quality improvement that to cause simple independent-qubit circuits to spin for the entire allowed 3min (or whatever it is) at O3.

alexanderivrii replied:
I would suggest to decrease the ``call_limit`` rather than simply returning ``None``, however I agree with leaving this as it is and looking into this in follow-up.

alexanderivrii reviewed (commented):
It is very satisfying to see how the previous modifications to `vf2_layout` allow to rewrite the `VF2PostLayout` pass in a much-much shorter and clearer way.

jakelishman committed:
Remove unnecesasry Debug bound

jakelishman committed:
Remove unused non-public `build_interaction_graph`

jakelishman committed:
Fix typo

jakelishman committed:
Fix incorrect exception message

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into vf2/prune/7

alexanderivrii reviewed (approved):
Thanks for doing the changes, and I absolutely love the revamped post-layout pass.

Cryoris commented:
Tagging this as `No changelog` for now, though this technically adds a C API function. We can still add that changelog for 2.3.0 final if we want.
